components/browser/icons/src/main/java/mozilla/components/browser/icons/extension/IconMessageHandler.kt

```diff
@@ -30,7 +30,16 @@ internal class IconMessageHandler(
 
     override fun onMessage(message: Any, source: EngineSession?): Any {
         if (message is JSONObject) {
-            message.toIconRequest()?.let { loadRequest(it) }
+            val messageRequest = message.toIconRequest()
+            val manifestRequest = session.webAppManifest?.toIconRequest()
+            val request = if (messageRequest != null && manifestRequest != null) {
+                // Merge the request resources
+                messageRequest.copy(resources = messageRequest.resources + manifestRequest.resources)
+            } else {
+                // Use the not-null request, if any
+                messageRequest ?: manifestRequest
+            }
+            loadRequest(request)
```

Comment:
This part I'm not sure about:
* It mixes the manifest icon into the list and we will try to load it and use it in the app UI. Do we actually need/want to do that? We can still load the manifest icon once we actually need it for an app launcher icon.
* There's now some race here: There's no guarantee that the manifest will be set on the `Session` by the time we receive this message from the icons add-on.

Reply:
I think we do want to mix in the manifest icons. The manifest represents replacing apple icons with a standard spec, so if we want to display high res apple icons we should want to display high res manifest icons.

Reply:
I don't have an idea yet how to deal with the race condition.

Reply:
Can we split this into a separate issue and look at that isolated (using manifest icons during normal icon load for a URL)? :)

Reply:
Filed #3642